python/python/bystro/covariance/hypothesis_hd.py

```diff
@@ -0,0 +1,873 @@
+
+    Returns
+    -------
+    k_value : float
+        Output of the kernel function.
+    """
+    if abs(x) >= 1.05:
+        return 0
+    if abs(x) <= 1:
+        return 1
+    return np.exp(1 / 0.05**2 - 1 / (0.05**2 - (abs(x) - 1) ** 2))
+
+
+def t_func(lambda_: np.ndarray, gamma: float, eta0: float) -> float:
+    """
+    Compute T value.
+
+    Parameters
+    ----------
+    lambda_ : np.ndarray
+        Array of eigenvalues.
+    gamma : float
+        Gamma value.
+    eta0 : float
+        Eta value.
+
+    Returns
+    -------
+    t_value : float
+        T value.
+    """
+    return sum(
+        (lambda_i - gamma) / eta0 * k_func((lambda_i - gamma) / eta0)
+        for lambda_i in lambda_
+    )
+
+
+def check_efficient(
+    gamma: float,
+    lambda1: np.ndarray,
+    lambda2: np.ndarray,
+    epsilon: float = 0.05,
+) -> bool:
+    """
+    Check if the splitting is efficient.
+    Parameters
+    ----------
+    gamma : float
+        Gamma value.
+    lambda1 : np.ndarray
+        First array of eigenvalues.
+    lambda2 : np.ndarray
+        Second array of eigenvalues.
+    epsilon : float, optional
+        Tolerance value, by default 0.05.
+    Returns
+    -------
+    is_efficient : bool
+        True if efficient, False otherwise.
+    """
+    range1 = abs(lambda1[0] - lambda1[-1])
+    range2 = abs(lambda2[0] - lambda2[-1])
+    return not (
+        max(abs(gamma - lambda1[0]), abs(gamma - lambda1[-1]))
+        > range1 - epsilon
+        or max(abs(gamma - lambda2[0]), abs(gamma - lambda2[-1]))
+        > range2 - epsilon
+    )
+
+
+def two_sample_test_(
+    X: np.ndarray,
+    Y: np.ndarray,
+    n: int,
+    const: float = 0.5,
+    alpha: float = 0.05,
+    epsilon: float = 0.05,
+    thres: Optional[float] = None,
+    mode: str = "test",
+    seed: int = 2021,
+) -> Union[dict, float]:
+    """
+    Perform a two-sample test.
+
+    Parameters
+    ----------
+    X : np.ndarray
+        First matrix.
+    Y : np.ndarray
+        Second matrix.
+    n : int
+        Sample size.
+    const : float, optional
+        Constant value, by default 0.5.
+    alpha : float, optional
+        Significance level, by default 0.05.
+    epsilon : float, optional
+        Tolerance value, by default 0.05.
+    thres : Optional[float], optional
+        Threshold value, by default None.
+    mode : str, optional
+        Mode of the test, by default "test".
+    seed : int, optional
+        Random seed, by default 2021.
+
+    Returns
+    -------
+    test_result : dict
+        Result of the two-sample test.
+    """
+    sample_list = two_way_sampling(X, Y, n, seed=seed)
+    Xs, Ys, Zs = sample_list["Xs"], sample_list["Ys"], sample_list["Zs"]
+
+    eig_xs = cov_eigs(Xs)
+    eig_ys = cov_eigs(Ys)
+    eig_zs = cov_eigs(Zs)
+    gamma = np.median(eig_zs)
+    if not check_efficient(gamma, eig_xs, eig_ys, epsilon):
+        return {"efficient": False, "c": 1}
+
+    eta0 = np.std(eig_zs, ddof=1) * const
```

Comment:
Could we also validate that const, alpha, epsilon, thres are in valid ranges? Typically we have been raising ValueError if an input is incorrect (e.g., `r("Invalid input: const must be a float that is greater than x")`